src/clusterfuzz/_internal/tests/core/bot/tasks/update_task_test.py

```diff
@@ -181,3 +184,79 @@ def test_get_source_url_linux(self):
+  def setUp(self):
+    helpers.patch_environ(self)
+    if os.getenv('PARALLEL_TESTS'):
+      self.skipTest('Package testing is disabled when running in parallel.')
+
+    self.temp_directory = self._make_temp_dir()
+    self.bot_tmpdir = self._make_temp_dir()
+    os.environ['ROOT_DIR'] = os.path.join(self.temp_directory, 'child')
+    os.mkdir(os.environ['ROOT_DIR'])
+    os.environ['BOT_TMPDIR'] = self.bot_tmpdir
+    os.environ['TEST_TMPDIR'] = self.bot_tmpdir
+    self.saved_cwd = os.getcwd()
+    os.chdir(os.environ['ROOT_DIR'])
+    helpers.patch(self, [
+        'clusterfuzz._internal.system.process_handler.cleanup_stale_processes',
+        'local.butler.common.is_git_dirty',
+        'clusterfuzz._internal.bot.tasks.update_task.get_source_url',
+        'clusterfuzz._internal.base.utils.read_data_from_file',
+    ])
+    self.mock.get_source_url.return_value = 'gs://clusterfuzz-deployment/test-deployment/linux-3.zip'
+    self.mock.read_data_from_file.return_value = b'version'
+
+  def tearDown(self):
+    os.chdir(self.saved_cwd)
+
+  def _make_temp_dir(self):
+    if not hasattr(self, '_dirs'):
+      self._dirs = []
+    created_dir = tempfile.TemporaryDirectory()
+    dir_name = created_dir.name
+    self._dirs.append(created_dir)
+    return dir_name
+
+  def test_files_have_read_and_write_permissions(self):
+    """Tests that all the extracted files have both read and write permissions."""
+    update_task.update_source_code()
+    for dirpath, _, filenames in os.walk(
+        os.path.join(self.temp_directory, 'clusterfuzz')):
+      for filename in filenames:
+        filepath = os.path.join(dirpath, filename)
+        self.assertTrue(os.access(filepath, os.R_OK))
+        self.assertTrue(os.access(filepath, os.W_OK))
+
+  def test_all_files_are_correctly_unpacked(self):
+    """Tests that all files in the zip archive are correctly unpacked."""
+    update_task.update_source_code()
+    zipfile_path = os.path.join(self._make_temp_dir(), 'linux.zip')
+    storage.copy_file_from(
+        'gs://clusterfuzz-deployment/test-deployment/linux-3.zip', zipfile_path)
+    archive = zipfile.ZipFile(zipfile_path)
+    for file in archive.namelist():
+      if os.path.basename(file) == 'adb':
+        continue
+      self.assertTrue(os.path.exists(os.path.join(self.temp_directory, file)))
+
+  def test_archive_execute_permission_is_respected(self):
+    """Tests that the exectuable bit is correctly propagated to source files."""
+    update_task.update_source_code()
+    zipfile_path = os.path.join(self._make_temp_dir(), 'linux.zip')
+    storage.copy_file_from(
+        'gs://clusterfuzz-deployment/test-deployment/linux-3.zip', zipfile_path)
+    archive = zipfile.ZipFile(zipfile_path)
+    for member in archive.infolist():
+      if os.path.basename(member.filename) == 'adb':
+        continue
+      mode = (member.external_attr >> 16) & 0o7777
+      filepath = os.path.join(self.temp_directory, member.filename)
+      if mode & 0o100:
```

Comment:
Should they also be readable too?

Reply:
This is already tested in another test above. All files are both readable and writable. This only ensures that the ones that are tagged as executable are actually executable.